Show constant value of binary operators in QuickInfo

Neme12 opened the pull request:
Fixes #26067 
Fixes #30951
Fixes #30960

Screenshots:
![quickinfoconstantvalue2](https://user-images.githubusercontent.com/11444821/46892779-e3e59b00-ce6e-11e8-831e-1245b0f60742.png)
![quickinfoconstantvalue1](https://user-images.githubusercontent.com/11444821/46892780-e5af5e80-ce6e-11e8-8be5-c2e542b40470.png)
![quickinfoconstantvalue3](https://user-images.githubusercontent.com/11444821/46892782-e6e08b80-ce6e-11e8-8e8a-ec33374de27f.png)
![quickinfoconstantvalue4](https://user-images.githubusercontent.com/11444821/46892785-e8aa4f00-ce6e-11e8-87d2-10f605d2cff9.png)
![quickinfoconstantvalue5](https://user-images.githubusercontent.com/11444821/46892786-e9db7c00-ce6e-11e8-9452-b5dd56c396a3.png)
![quickinfoconstantvalue6](https://user-images.githubusercontent.com/11444821/46892788-eba53f80-ce6e-11e8-87a0-2966937ce8a2.png)
![quickinfoconstantvaluelongstr](https://user-images.githubusercontent.com/11444821/46892800-f2cc4d80-ce6e-11e8-89a8-caa9aa980b66.png)

@sharwell Can you please take this to a design review? Thanks.

sharwell commented:
Additional test cases:

1. Cases where the value assigned to a `const` field is not actually constant. This is a compilation error, but the results should still be sane.
2. Cast operator used on constant values, e.g. hovering over `==` in `(byte)3 == (byte)4`.

CyrusNajmabadi commented:
I think i'd prefer if the constant value were just printed, and not the full equation again.  I'm finding those hard to parse.

i.e. it looks like:  `Constant Value: 5 + 10 = 15`, which makes my brain think "the constant value is an assignment?"

CyrusNajmabadi commented on src/Features/Core/Portable/QuickInfo/CommonSemanticQuickInfoProvider.cs:
```diff
@@ -296,6 +303,92 @@ void AddSection(string kind, ImmutableArray<TaggedText> taggedParts)
+            if (!syntaxFacts.IsBinaryExpression(token.Parent))
+            {
+                return default;
+            }
```
why not any expression that can be constant?

Neme12 replied:
Because the issue asked for just binary expressions. I would prefer to have this on any expression too, I'll wait for the design review though.

CyrusNajmabadi replied:
i'd say that things don't need to be terribly literal (no pun intended) :)

Neme12 replied:
@CyrusNajmabadi I asked what kinds of other expressions this should show up in with an extensive comment below, and you said none of them. So it's unclear to me what you were requesting here.

CyrusNajmabadi replied:
i'm ok with this behavior.

CyrusNajmabadi commented on src/Features/Core/Portable/QuickInfo/CommonSemanticQuickInfoProvider.cs:
```diff
@@ -296,6 +303,92 @@ void AddSection(string kind, ImmutableArray<TaggedText> taggedParts)
+
+            void AddConstantText(object value)
+            {
+                textBuilder.AddRange(TrimTaggedTextRun(LiteralDisplayBuilder.Build(syntaxFacts, generator, value), 42));
```
why 42?

Neme12 replied:
It looked just right.

Neme12 replied:
If we end up not showing the binary operator with both operands and only the result, it may be appropriate to make this larger.

CyrusNajmabadi commented on src/Features/Core/Portable/QuickInfo/LiteralDisplayBuilder.cs:
```diff
@@ -0,0 +1,70 @@
+                    ? TextTags.Keyword :
+                syntaxFacts.IsNumericLiteralExpression(node)
+                    ? TextTags.NumericLiteral :
+                TextTags.Text;
```
i don't know what's going on here at all :)

Neme12 replied:
This is a WIP. I will add comments later, after the design is finished.

Neme12 replied:
➡️ This code no longer exists.

CyrusNajmabadi commented on src/Features/Core/Portable/QuickInfo/LiteralDisplayBuilder.cs:
```diff
@@ -0,0 +1,70 @@
+                    ? TextTags.NumericLiteral :
+                TextTags.Text;
+
+            textBuilder.Add(new TaggedText(tag, node.ToString()));
```
strings can be multiline.  what's our behavior here?

Neme12 replied:
For C# it is `"First Line\nSecond Line"`
For VB it is `"First Line"&vbCrLf&"Second Line"`

CyrusNajmabadi commented on src/Features/Core/Portable/QuickInfo/LiteralDisplayBuilder.cs:
```diff
@@ -0,0 +1,70 @@
+        {
+            var textBuilder = new List<TaggedText>();
+            AddConcatenations(textBuilder, syntaxFacts, generator.LiteralExpression(value));
+            return textBuilder.ToImmutableArray();
```
ArrayBuilder?

CyrusNajmabadi commented on src/Features/Core/Portable/QuickInfo/LiteralDisplayBuilder.cs:
```diff
@@ -0,0 +1,70 @@
+                node = syntaxFacts.GetNameOfMemberAccessExpression(node);
+            }
+
+            return node;
```
Not really getting what this is for.  Are there tests for this?

Neme12 replied:
This is all for VB, which can't represent certain characters (such as a line break) in a string literal, instead their generated as concatenations with constants. This simplifies the member access of those constants so that we don't show the full namespace or whatever it is contained in.

CyrusNajmabadi replied:
Interesting.  And definitely worth a comment in the code to help make that clear :)

Neme12 replied:
➡️ This code no longer exists.

Neme12 commented:
> I think i'd prefer if the constant value were just printed, and not the full equation again. I'm finding those really hard to parse.
> 
> i.e. it looks like: `Constant Value: 5 + 10 = 15`, which makes my brain think either"the constant value is an assignment?" or "the constant value is `5 + 10 == 15`? isn't that just 'true'?"
> 
> Basically, if i need the subparts, i can get that by hovering. If i want the final value, it would be nice to have it as is IMO.

I tend to agree. I'll wait for the design review to make a decision.

CyrusNajmabadi commented:
> I tend to agree. I'll wait for the design review to make a decision.

Sure... though as the person creating the PR, i'd say the design is very much in your hands.  It's not like this original suggestion is something that has to be followed to the letter.  You can use your best judgement to create the best feature you think you can, and then get feedback/design-reviews on that. 

In other words, the design review is generally useful when it's *necessary* to get uncertain things decided on. It's something that can, and definitely should, be side-stepped if not necessary :)

Neme12 commented:
@sharwell I have at least 2 questions for the design review:
1. Should we reconstruct the binary operator (like the screenshots show now) or just show the result? I agree with @CyrusNajmabadi that it seems a little bit out of place, especially in cases like `a == b = true` or in VB: `a = b = true`. It's an odd mix of mathematical notation and a programming language.
2. Should we show this on all expressions, as opposed to just binary operators? Normally, I would just make it so that we check the parent node of the token and if it's an expression (probably except for literals), show its constant value (if there is one). That would show this on identifiers, all operators, also on default expressions, and maybe other things that I can't think of now. The only concern is that with identifiers for constants, their definition is already shown in quick info, so we might show the same thing twice:
![image](https://user-images.githubusercontent.com/11444821/46905392-35d00480-cef3-11e8-950c-c3ea34643bc2.png)
On the other hand, the constant might be defined or computed from other constants (not just with a literal) and in those cases it would be very useful to see the constant value, for example:
![image](https://user-images.githubusercontent.com/11444821/46905453-4b91f980-cef4-11e8-9d5d-ebd171532faa.png)

CyrusNajmabadi commented:
> The only concern is that with identifiers for constants, their definition is already shown in quick info, so we might show the same thing twice:

In that case, i would say there's no real benefit :)  If it's already a constant symbol, don't bother.  However, if it's a non-trivial expression (i.e. not a single token) that produces a constant value, then show it.

Neme12 commented:
@CyrusNajmabadi But even though it's a constant symbol, it might be initialized with a complex expression. Quick info only shows the complex expression, not it's value. So yes there would still be benefit.

CyrusNajmabadi commented:
> @CyrusNajmabadi But even though it's a constant symbol, it might be initialized with a complex expression. Quick info only shows the complex expression, not it's value. So yes there would still be benefit.

Maybe :)

Neme12 commented:
I changed the implementation strategy from generating expressions to using a new added compiler API: `SymbolDisplay.FormatPrimitiveToDisplayParts`.

I believe this is the right approach because:
1. Generating an expression that we then want to show the user would require simplification. That is not trivial because it's needed to distinguish cases like `vbCrLf` (which is a field member access that has to be simplified all the way down to the name) and `int.MaxValue`, where we have to simplify `System.Int32` to the predefined type. This could still be done (but we'd be duplicating what the compiler already does).
2. The previous approach generated things like `int.MaxValue` instead of the actual value. While special fields like this are useful for generating code, I don't believe this is what we want to show in quick info where we'd like to know what the actual value is. There's an option in the expression generator to disable this, but then we'd generate things like `0 / 0` for `double.NaN`.

I also think what is displayed here should match what quick info already does for constant values of fields or default parameters, which is calling ToDisplayParts and using the compiler implementation. The new API I added is mostly using the existing implementation. This way, all constant values in quick info should be displayed in a consistent way.

Neme12 commented on src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt:
```diff
@@ -142,6 +142,7 @@ override Microsoft.CodeAnalysis.CSharp.Syntax.NullableDirectiveTriviaSyntax.IsAc
 override Microsoft.CodeAnalysis.CSharp.Syntax.RangeExpressionSyntax.Accept(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor visitor) -> void
 override Microsoft.CodeAnalysis.CSharp.Syntax.RangeExpressionSyntax.Accept<TResult>(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor<TResult> visitor) -> TResult
 static Microsoft.CodeAnalysis.CSharp.LanguageVersionFacts.TryParse(string version, out Microsoft.CodeAnalysis.CSharp.LanguageVersion result) -> bool
+static Microsoft.CodeAnalysis.CSharp.SymbolDisplay.FormatPrimitiveToDisplayParts(object obj, bool quoteStrings, bool useHexadecimalNumbers) -> System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.SymbolDisplayPart>
```
:pencil: Parameters are modeled after the existing `FormatPrimitive` method.

Neme12 commented on src/Compilers/Core/Portable/SymbolDisplay/ObjectDisplayOptions.cs:
```diff
@@ -40,5 +40,11 @@ internal enum ObjectDisplayOptions
+        /// In C# when specifying <see cref="IncludeCodePoints"/>, use a hexadecimal number for the code point.
+        /// In Visual Basic, use hexadecimal numbers in calls to ChrW.
+        /// </summary>
+        UseHexadecimalNumbersForCharacters = 1 << 5
```
The existing `FormatPrimitive` methods use hexadecimal numbers for both numeric literals and character numbers based on the `useHexadecimalNumbers` parameter, whereas the existing implementation for constants of fields and default parameters inside `ToDisplayParts` always uses decimal numeric literals and hexadecimal numbers inside `ChrW`. To accomodate for this flexibility, I added this option.

I do find it slightly unfortunate that the new API won't have any option to behave the same way as constants displayed for symbols. This means that quick info will be inconsistent after all (showing hexadecimal for constant fields and default parameters but always decimal for results of expressions), but I don't know what can/should be done about it. I don't care that much about special characters in VB.

CyrusNajmabadi replied:
Why can't this be controlled using SymbolDisplayOptions?

Neme12 replied:
Maybe. I'll wait for compiler review to see what their preference is.

Neme12 replied:
@CyrusNajmabadi That sounds like a good idea. We could maybe put this inside `SymbolDisplayMiscellaneousOptions` or make a new enum: `SymbolDisplayConstantValueOptions`. We could also add an overload of `FormatPrimitive` taking this. Although we'd need to figure out how to make it so that 2 of those options are on by default to match the existing default behavior for symbol display.

CyrusNajmabadi replied:
> SymbolDisplayConstantValueOptions 

That's my preference.

> Alternatively we could negate them

That's my preference as well.  The quotes one, in particular, should be a 'NoQuotes' value.  By default, strings should be printed as they would be as literals (i.e. with quotes).  So you'd the option to opt out of that instead of needing to opt into it.

Neme12 replied:
@CyrusNajmabadi There's also the problem that for regular numeric literals, decimal should be the default (so there would be an option like UseHexadecimalNumbers) but for numbers for character values, the default should be the opposite to match current behavior. So then we'd need a strange option like DoNotUseHexadecimalNumbersForEscapedCharacters :(

Neme12 replied:
or maybe UseDecimalNumbersForCharacters?

CyrusNajmabadi replied:
can UseDecimal be the default, and have UseHexa be something that someone needs to opt into?  Existing code that goes through FormatPrimitive will map their booleans to teh right symbol options.

Neme12 commented on src/EditorFeatures/TestUtilities/QuickInfo/AbstractSemanticQuickInfoSourceTests.cs:
```diff
@@ -49,52 +34,66 @@ internal Action<QuickInfoItem> WarningGlyph(Glyph expectedGlyph)
-        protected void AssertTaggedText(
-            string expectedText,
-            ImmutableArray<TaggedText> taggedText,
-            FormattedClassification[] expectedClassifications = null)
```
The `expectedClassifications` parameter was unused and existing tests that used it were ineffectual.

CyrusNajmabadi commented:
> The previous approach generated things like int.MaxValue instead of the actual value. While special fields like this are useful for generating code, I don't believe this is what we want to show in quick info, where we'd like to know what the actual value is. There's an option in the expression generator to disable this, but then we'd generate strange things like 0 / 0 for double.NaN instead of NaN.

I'm not sure i actually agree with this.  if some computation evaluates out to 'double.NaN', then just show me 'double.NaN' :)  Similar to int.MaxValue. if i've done a bunch of constant operations on ints and it ended up that i have int.MaxValue, it's super nice to know that (versus some number like 2147483647 which i have to remember is that value).  That said, i'm not going to press hard on this.  Both seem reasonable.  But i do think there's an argument we want the special values for QI.

Neme12 commented on src/EditorFeatures/VisualBasicTest/QuickInfo/SemanticQuickInfoSourceTests.vb:
```diff
@@ -106,25 +106,25 @@ Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.UnitTests.QuickInfo
+                    ("System", TextTags.Namespace),
+                    (".", TextTags.Operator),
+                    ("Int32", TextTags.Struct)
+                ))
```
As I mentioned above, the previous `expectedClassifications` argument was unused and not actually tested. I instead added a way to verify the `TextTag`s and rewrote existing tests to use it. I believe testing `TextTag` is the right thing to do because that's what's returned by the actual public quick info API, and converting it to classifications loses some information (for example locals & parameters, fields, methods are all identifier).

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor_Constants.cs:
```diff
@@ -1,59 +1,40 @@
         }
 
-        protected override void AddLiteralValue(SpecialType type, object value)
+        private static bool CanDefinitelyBeNull(ITypeSymbol type)
```
Maybe 'CanUseNullLiteral'?

Neme12 replied:
Can definitely be used... :smile: my idea with "definitely" was to make it clear that this might not handle all cases - it doesn't return true for constrained type parameters. The important part is that when it does, we can use null.

CyrusNajmabadi replied:
Gotcha :)  i'm fine either way.  Thanks!

Neme12 replied:
➡️ Renamed in 5e47bdaf56331c8ea0214ad9afe0f3848fc255f0

Neme12 commented:
> I'm not sure i actually agree with this. if some computation evaluates out to 'double.NaN', then just show me 'double.NaN' :) Similar to int.MaxValue. 

For `double.NaN`, the result is `NaN`. If you feel like boundary values of integer types should be shown as special fields, that's fine, but I think it should be consistent with what quick info shows on default parameters and constant fields, which is to show the integer value in these cases. We could probably add a new option to the compiler API to support this, although I don't really feel like wanting to do that.

CyrusNajmabadi commented:
> For double.NaN, the result is NaN. If you feel like boundary values of integer types should be shown as special fields, that's fine, but I think it should be consistent with what quick info shows on default parameters and constant fields, which is to show the integer value in these cases.

I see the following:

![image](https://user-images.githubusercontent.com/4564579/48023974-2f007000-e0f4-11e8-88e7-24af55537642.png)

Neme12 commented:
@CyrusNajmabadi That is because it's in source, in which case it doesn't use the compiler API and instead shows literally the syntax you have.

Neme12 commented:
> but this is what quick-info shows.

Sure, that is what quick info shows if that's the syntax you have. But in case of showing constant values of arbitrary binary expressions, there is no existing syntax to go by. For all we know, it might be something like `2147483646 + 1`.

Neme12 commented:
retest vsi please

CyrusNajmabadi commented:
IMO, it's appropriate to show constants when it's non-obvious what they are.  So, for example, showing constants for things like 'default' doesn't seem that helpful.  Telling people `default(bool)` is `false` just seems like it's more about showing we *can* do it, versus any amount of particular value for a user.

In terms of the list:

> Do we show this on unary operators? My opinion is that it wouldn't be very useful (but we can if it doesn't hurt to show extra info).

My expectation is that much of the time people will care about the constant value of an actual symbol.  So there will be a name to hover over.  Needing to understand expression *minutia* seems less relevant.  

> Do we show this on the ternary operator - ? and : tokens?

See above.  Seems too niche to be that valuable.

> On literals? At a first glance it looks like this wouldn't be helpful at all since we see the value in the literal itself, but this might be handy for binary & hex literals. So do we always show it on literals? Or never? Or only on non-decimal literals?

All getting too niche.  I don't think the goal is to be able to provide every shred of data in every event "just in case" there is some hypothetical person who might find it valuable.  Instead, i think the information needs to be highly relevant much of the time and has to provide enough value to warrant the UI inclusion cost.

in other words, this is similar to language design.  We start at -100, and we need enough value to push it above that.

--

Note: for most of the above, i would not actively block things.  However, at the same time, i would say we shouldn't put in extra effort to support these things.

Neme12 commented:
@CyrusNajmabadi so you'd prefer nothing out of that list? only binary expressions?

Neme12 commented on src/Compilers/Core/Portable/SymbolDisplay/AbstractSymbolDisplayVisitor.cs:
```diff
@@ -71,27 +72,33 @@ protected AbstractSymbolDisplayVisitor NotFirstVisitorNamespaceOrType
+        protected const ObjectDisplayOptions LiteralDisplayOptions =
+            ObjectDisplayOptions.UseQuotes |
+            ObjectDisplayOptions.EscapeNonPrintableCharacters |
+            ObjectDisplayOptions.UseHexadecimalNumbersForCharacters;
```
Done.

Neme12 commented:
@CyrusNajmabadi I updated the compiler APIs with new options as we discussed. Could you take another look at them? Thanks. (IDE portion stayed the same).

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplay.cs:
```diff
@@ -189,5 +180,91 @@ public static string FormatLiteral(char c, bool quote)
+        /// <see cref="byte"/>, <see cref="short"/>, <see cref="ushort"/>, <see cref="int"/>, <see cref="uint"/>,
+        /// <see cref="long"/>, <see cref="ulong"/>, <see cref="double"/>, <see cref="float"/>, <see cref="decimal"/>,
+        /// and <see langword="null"/>.
+        /// </remarks>
```
1. just to check, but it's intentional that dates are not in this list?
2. this list doesn't mention enums, but i can see that is handled in teh check below

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplay.cs:
```diff
@@ -189,5 +180,91 @@ public static string FormatLiteral(char c, bool quote)
+            else
+            {
+                builder.Add(new SymbolDisplayPart(SymbolDisplayPartKind.Keyword, null, SyntaxFacts.GetText(SyntaxKind.NullKeyword)));
+            }
```
nit, consider inverting this 'if'.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplay.cs:
```diff
@@ -189,5 +180,91 @@ public static string FormatLiteral(char c, bool quote)
+            }
+        }
+
+        private static void AddLiteralValue(ArrayBuilder<SymbolDisplayPart> builder, object value, ObjectDisplayOptions options)
```
consider: AddNonNullLiteralValue.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor_Constants.cs:
```diff
@@ -1,59 +1,40 @@
+            }
+            else
+            {
+                SymbolDisplay.AddConstantValue(builder, value, format.ConstantValueOptions);
```
based on this logic flow, it looks like this guy is handling the null case, yes?

Neme12 replied:
Yes.

CyrusNajmabadi commented on src/Compilers/Core/Portable/SymbolDisplay/NumericFormat.cs:
```diff
@@ -0,0 +1,10 @@
+    public enum NumericFormat
+    {
+        Decimal,
+        Hexadecimal
```
public API, so doc. maybe with examples.

CyrusNajmabadi commented on src/Compilers/Core/Portable/SymbolDisplay/ObjectDisplayOptions.cs:
```diff
@@ -40,5 +40,11 @@ internal enum ObjectDisplayOptions
+
+        /// <summary>
+        /// In C# when specifying <see cref="IncludeCodePoints"/>, use a hexadecimal number for the code point.
+        /// In Visual Basic, use hexadecimal numbers in calls to ChrW.
```
miht be worth calling out what happens if this isn't set. (or don't, since it seems reasonably obvious)

CyrusNajmabadi commented on src/Compilers/Core/Portable/SymbolDisplay/SymbolDisplayConstantValueOptions.cs:
```diff
@@ -0,0 +1,38 @@
+        public NumericFormat NumericLiteralFormat { get; }
+
+        /// <summary>
+        /// For Visual Basic, specifies how numeric values of characters inside ChrW are displayed.
```
oh, does thsi not affect C# at all?  do we not use \uXXXX chars there if htis is set?

Neme12 replied:
C# only has hex escape sequence. We always use `\uXXXX`.

CyrusNajmabadi commented on src/Compilers/Core/Portable/SymbolDisplay/SymbolDisplayFormat.cs:
```diff
@@ -368,7 +379,39 @@ public class SymbolDisplayFormat
             SymbolDisplayKindOptions kindOptions = default(SymbolDisplayKindOptions),
-            SymbolDisplayMiscellaneousOptions miscellaneousOptions = default(SymbolDisplayMiscellaneousOptions))
+            SymbolDisplayMiscellaneousOptions miscellaneousOptions = default(SymbolDisplayMiscellaneousOptions),
+            SymbolDisplayConstantValueOptions constantValueOptions = default)
```
can you use 'default' consistently here?  either all 'default', or all 'default(...)'

CyrusNajmabadi commented on src/Compilers/Core/Portable/SymbolDisplay/SymbolDisplayFormat.cs:
```diff
@@ -400,7 +444,8 @@ public class SymbolDisplayFormat
             SymbolDisplayKindOptions kindOptions = default(SymbolDisplayKindOptions),
-            SymbolDisplayMiscellaneousOptions miscellaneousOptions = default(SymbolDisplayMiscellaneousOptions))
+            SymbolDisplayMiscellaneousOptions miscellaneousOptions = default(SymbolDisplayMiscellaneousOptions),
+            SymbolDisplayConstantValueOptions constantValueOptions = default)
```
same.

CyrusNajmabadi commented on src/Compilers/Core/Portable/SymbolDisplay/SymbolDisplayFormat.cs:
```diff
@@ -753,6 +806,31 @@ public SymbolDisplayFormat RemoveLocalOptions(SymbolDisplayLocalOptions options)
+                this.LocalOptions,
+                this.KindOptions,
+                this.MiscellaneousOptions,
+                options);
```
wow.  :)  why isn't this type using the standard 'With' impl.  that would make this code into simply: `return this.With(constantVallueOptions: options)` and would make all the other methods so much shorter.

CyrusNajmabadi commented on src/Compilers/VisualBasic/Portable/SymbolDisplay/ObjectDisplay.vb:
```diff
@@ -333,7 +333,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.ObjectDisplay
-        Private Function Identifier(c As Char) As Integer
+        Private Function FieldName(c As Char) As Integer
+            Return (SymbolDisplayPartKind.FieldName << 16) Or AscW(c)
+        End Function
```
i'm not really sure what's going on here.

Neme12 replied:
I don't know why this code is so complex, but I'd prefer to keep it that way.

CyrusNajmabadi commented on src/Compilers/VisualBasic/Portable/SymbolDisplay/ObjectDisplay.vb:
```diff
@@ -420,13 +424,13 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.ObjectDisplay
                         If wellKnown IsNot Nothing Then
                             For Each e In wellKnown
-                                Yield Identifier(e)
+                                Yield FieldName(e)
```
not sure here either.

CyrusNajmabadi replied:
i think that's worth a comment :)

CyrusNajmabadi commented on src/Compilers/VisualBasic/Portable/SymbolDisplay/SymbolDisplay.vb:
```diff
@@ -156,24 +204,52 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+            End If
+
+            If Not constantValueOptions.NoQuotes Then
+                options = options Or ObjectDisplayOptions.UseQuotes Or ObjectDisplayOptions.EscapeNonPrintableCharacters
```
interesting that this covers two options.  can you doc why?

Neme12 replied:
I don't know why. I am just preserving the existing behavior of the `useQuotes` flag.

CyrusNajmabadi replied:
that's worth a comment :)

Neme12 replied:
Why is it worth a comment to state my personal ignorance? I'm sure other people do know.

CyrusNajmabadi replied:
because it helps document to the next person.  Consider someone else coming along and seeing this and going "hrmm.. was this intentional?  why is this going on?"  Perhaps they're doing it to try to fix some bug.  Having a comment explaining that this was just how the previous system worked, and that we're preserving it, can be helpful.

Neme12 replied:
But that's the thing. I do *not* know whether it's intentional. I am the "someone else coming along" that you're describing.

CyrusNajmabadi commented on src/Features/Core/Portable/QuickInfo/CommonSemanticQuickInfoProvider.cs:
```diff
@@ -296,6 +304,91 @@ void AddSection(string kind, ImmutableArray<TaggedText> taggedParts)
+                return default;
+            }
+
+            var textBuilder = ImmutableArray.CreateBuilder<TaggedText>();
```
i still would not do this.

Neme12 replied:
Would not do what? Use an array builder?

CyrusNajmabadi replied:
i would not piece together an expression as the result.  I would just take the final constant value for the binary expressoin and show that.  As a bonus, if you are dealing with expressions with large constants,there's less that needs to be trimmed.

Neme12 replied:
Ask @sharwell about that, I am not in charge of the design.

CyrusNajmabadi commented on src/Features/Core/Portable/QuickInfo/CommonSemanticQuickInfoProvider.cs:
```diff
@@ -296,6 +304,91 @@ void AddSection(string kind, ImmutableArray<TaggedText> taggedParts)
+            ImmutableArray<TaggedText> FormatValue(object value)
+            {
+                var taggedText = displayService.FormatPrimitiveToDisplayParts(value, options: default).ToTaggedText();
+                return TrimTaggedTextRun(taggedText, 42);
```
not a fan of random constants.

CyrusNajmabadi commented on src/Features/Core/Portable/QuickInfo/CommonSemanticQuickInfoProvider.cs:
```diff
@@ -296,6 +304,91 @@ void AddSection(string kind, ImmutableArray<TaggedText> taggedParts)
+
+        private static ImmutableArray<TaggedText> TrimTaggedTextRun(ImmutableArray<TaggedText> taggedTextRun, int maxLength)
+        {
+            const string UnicodeEllipsis = "\u2026";
```
feels like it would be better to me if we just showed the final constant value.

Neme12 replied:
Ask @sharwell, he's been silent about that (and it seems it wasn't brought up in the design review).

CyrusNajmabadi replied:
Putting on my IDE designer hat: there is value in showing the constant value.  Let's *start* with that and see if there's really a compelling need to do more.  Absent that, we should just go with something simpler.

Neme12 replied:
You do not need to convince me - I agree that would be preferrable here. You need to convince @sharwell.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Classification/ClassificationHelpers.cs:
```diff
@@ -1,16 +1,21 @@
 {
-    internal static class ClassificationHelpers
+    [ExportLanguageService(typeof(IClassificationHelpersService), LanguageNames.CSharp), Shared]
+    internal sealed class ClassificationHelpers : IClassificationHelpersService
```
who imports this service?

Neme12 replied:
Nobody, I forgot to revert this. Thanks!

Neme12 replied:
➡️ Fixed in fc20d2c6da76740bc23ce9ff1c012fd87fd29426

sharwell commented:
@Neme12 I'll try to gather up the outstanding questions here and bring them for a review Monday

sharwell commented:
> All these cases are assuming that the expression has a constant value:
> 1. Do we show this on unary operators? My opinion is that it wouldn't be very useful (but we can if it doesn't hurt to show extra info).

We agree (does not hurt). If support for these operators is not already implemented, we would not block on it.

> 2. Do we show this on the ternary operator - `?` and `:` tokens? My preference would be: yes.

We agree (yes).

> 3. On the `default` keyword for default literals and default expressions? I think this would be nice (although maybe not very useful since the value should be clear from the type).

We agree (does not hurt).

> 4. On parenthesized expressions - the `(` and `)` tokens? This seems a little odd.
> 5. On casts - for `(` and `)` tokens? This would seem very odd.

We don't think parentheses tokens need to trigger the quick info.

> 6. On `checked` and `unchecked` keywords for checked/unchecked expressions? My opinion: I don't see why not.

We agree (could help, especially for `unchecked` expressions that contain a cast that loses precision).

> 7. On `nameof`? This might be nice too but at the same time not at all useful.

We agree (could help users who are not as familiar with `nameof`, e.g. when looking at `nameof(List<int>)`).

> 8. On literals? At a first glance it looks like this wouldn't be helpful at all since we see the value in the literal itself, but this might be handy for binary & hex literals. So do we always show it on literals? Or never? Or only on non-decimal literals?

We agree (helpful as an alternate-base presentation). It doesn't seem to hurt to show on decimal literals.

> 9. On variables that are themselves constant? This has been discussed before in that it might be redundant in most cases but there are some cases where it would be very useful if the initializer itself is a complex expression that doesn't show the value (this example is from Roslyn):
> ![image](https://user-images.githubusercontent.com/11444821/46905453-4b91f980-cef4-11e8-9d5d-ebd171532faa.png)

This seems very helpful, with your screenshot highlighting the case where it's helpful.

sharwell commented:
> Should we reconstruct the binary operator with an equal sign (like the screenshots show now) or just show the result?

Reconstruction seems like a good presentation, if possible. It answers several potential questions for a developer in a quick glance.

> Should we show this on all expressions, as opposed to just binary operators?

Most expressions would show the information, many of which are itemized in the previous reply.

Neme12 commented on src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs:
```diff
@@ -112,8 +112,8 @@ private async Task VerifyWithMscorlib45Async(string markup, Action<QuickInfoItem
-            await TestWithOptionsAsync(Options.Regular, markup, expectedResults);
-            await TestWithOptionsAsync(Options.Script, markup, expectedResults);
+            await TestWithOptionsAsync(Options.Regular.WithLanguageVersion(LanguageVersion.Latest), markup, expectedResults);
+            await TestWithOptionsAsync(Options.Script.WithLanguageVersion(LanguageVersion.Latest), markup, expectedResults);
```
This was necessary because of #30384

Neme12 commented on src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs:
```diff
@@ -1,267 +1,17 @@
 namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.QuickInfo
 {
-    public partial class SemanticQuickInfoSourceTests : AbstractSemanticQuickInfoSourceTests
+    public sealed class SemanticQuickInfoSourceTests : SemanticQuickInfoSourceTestsBase
```
The change in this commit (9d954cb697a34d565480bebb81f973035c287cf2) - separating parts of a test class into separate classes while extracting the common helpers into a base is a really good use case for #31837. I got a bunch of accessibility errors because I needed to change each of those helpers from private to protected and had to go through them one by one by hitting F12 and changing the modifier.

Neme12 replied:
This might even be a reason to have a fix all for that.

Neme12 commented:
@sharwell I think you should review the presentation for when there is a constant value but no associated symbol, like there usually is. It looks like this:
![image](https://user-images.githubusercontent.com/11444821/51498558-4122ae00-1dc7-11e9-8e16-58c3a5aaea9e.png)

RikkiGibson commented:
what's the status on this @Neme12 @sharwell?

sharwell commented:
@Neme12 I'm planning to move this forward as part of the 16.3 improvements to Quick Info. Since the timeline is somewhat time sensitive (our fault not yours) I'll go ahead and fix the merge conflicts and update the tests to use the new Quick Info test harness. I'll also plan to make any changes necessary to account for any design concerns we find. If you want to do these yourself let me know.

Neme12 commented:
@sharwell Thanks for the info. Please feel free to take over this PR.

sharwell commented:
@Neme12 thanks! I had to move this back to 16.4 because the compiler changes don't align well with our shipping schedule, but I'm still looking forward to moving it forward and getting to use it!

jcouv commented:
Marking as personal PR to hide from the compiler's code review backlog until API design is finalized offline.

sharwell commented:
@333fred can we schedule an API review for the @dotnet/roslyn-compiler side of this?

333fred commented:
@sharwell what timeframe are we talking about here?

jinujoseph commented:
@333fred  now for 16.6

333fred commented:
@Neme12 sorry for the delay on this, I missed Jinu's ping and then lost track of it completely. This is scheduled for review on Thursday afternoon.

sharwell commented:
I'm working on fixing the merge conflicts and updating this to account for recent changes.

333fred commented:
Notes from API review:

* We want to slim down the public api changes. This PR exposes a whole bunch of options and then uses the defaults for all of them. Let's make the minimal change here, and add more options in future overloads if we find use cases for them.
* We accept two new APIs on the `SymbolDisplay` classes: `PrimitiveToDisplayParts` and `PrimitiveToMinimalDisplayParts`
    * Takes an object constant and possible type symbol (for enums, if necessary)
    * `PrimitiveToMinimalDisplayParts` will take a position and SemanticModel to ensure that any enum names are appropriately qualified as necessary. `PrimitiveToDisplayParts` will not, and will result in fully-qualified names. This follows the same pattern as our existing `To(Minimal)DisplayParts` overloads.
    * No options will exposed. If we find that we need more options at a later point, we'll add new overloads then.
* `SymbolDisplayConstantValueOptions` should remain internal only.

@sharwell will continue working with @Neme12 to adjust as necessary. When this is ready for review, please remove the PR for Personal Review Only label and we'll know to take a look. Thanks!

sharwell commented:
@333fred for initial review on the compiler side. Is 16.6 Preview 3 acceptable for this, or 16.7 Preview 1?

AlekseyTs commented on src/Compilers/Core/Portable/PublicAPI.Unshipped.txt:
```diff
@@ -1,2 +1,2 @@
 Microsoft.CodeAnalysis.CommandLineSourceFile.CommandLineSourceFile(string path, bool isScript, bool isInputRedirected) -> void
-Microsoft.CodeAnalysis.CommandLineSourceFile.IsInputRedirected.get -> bool
\ No newline at end of file
+Microsoft.CodeAnalysis.CommandLineSourceFile.IsInputRedirected.get -> bool
```
It looks like there are no meaningful changes in this file. Revert?

AlekseyTs commented on src/Compilers/Core/Portable/SymbolDisplay/AbstractSymbolDisplayVisitor.cs:
```diff
@@ -75,25 +76,31 @@ protected AbstractSymbolDisplayVisitor NotFirstVisitorNamespaceOrType
         {
-            Debug.Assert(constantValue != null);
+            Debug.Assert(value is null || value.GetType().IsPrimitive || value is string || value is decimal || value is DateTime);
+            if (value is object && type.OriginalDefinition.SpecialType == SpecialType.System_Nullable_T)
```
>type.OriginalDefinition.SpecialType == SpecialType.System_Nullable_T [](start = 35, length = 68)

Consider using nullable helpers from ITypeSymbolHelpers type

AlekseyTs commented on src/Compilers/Core/Portable/SymbolDisplay/AbstractSymbolDisplayVisitor.cs:
```diff
@@ -75,25 +76,31 @@ protected AbstractSymbolDisplayVisitor NotFirstVisitorNamespaceOrType
+            if (value is object && type.OriginalDefinition.SpecialType == SpecialType.System_Nullable_T)
+            {
+                // The non-null value will be an instance of the nullable underlying type
+                type = ((INamedTypeSymbol)type).TypeArguments[0];
```
>type = ((INamedTypeSymbol)type).TypeArguments[0]; [](start = 16, length = 49)

Here as well.

AlekseyTs commented:
>     /// Specifies the options for how generics are displayed in the description of a symbol.

Is this a typo?

---
Refers to: src/Compilers/Core/Portable/SymbolDisplay/ObjectDisplayOptions.cs:10 in 79aae9f. [](commit_id = 79aae9fce09d737fd794031f5f1d5b3bb4426fac, deletion_comment = False)

AlekseyTs commented:
>  Is 16.6 Preview 3 acceptable for this, or 16.7 Preview 1?

@agocke ?

AlekseyTs commented on src/Compilers/Core/Portable/SymbolDisplay/SymbolDisplayFormat.cs:
```diff
@@ -322,47 +323,16 @@ public class SymbolDisplayFormat
-        /// </param>
+        internal SymbolDisplayCompilerInternalOptions CompilerInternalOptions { get; }
+
         public SymbolDisplayFormat(
```
>SymbolDisplayFormat [](start = 15, length = 19)

Not sure why did we drop a doc comment for this method.

AlekseyTs commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplay.cs:
```diff
@@ -295,5 +286,146 @@ public static string FormatLiteral(char c, bool quote)
+        /// </remarks>
+        public static ImmutableArray<SymbolDisplayPart> PrimitiveToDisplayParts(ITypeSymbol? type, object? value, SymbolDisplayFormat? format = null)
+        {
+            if (!(value is null || value.GetType().IsPrimitive || value is string || value is decimal))
```
>(value is null || value.GetType().IsPrimitive || value is string || value is decimal) [](start = 17, length = 85)

Consider extracting and reusing this check

agocke commented:
@AlekseyTs Always fine to put into master, it'll go into the next appropriate release.

AlekseyTs commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplay.cs:
```diff
@@ -295,5 +286,146 @@ public static string FormatLiteral(char c, bool quote)
+
+        private static void AddLiteralValue(ArrayBuilder<SymbolDisplayPart> builder, object value, ObjectDisplayOptions options)
+        {
+            Debug.Assert(value.GetType().IsPrimitive || value.GetType().IsEnum || value is string || value is decimal);
```
>value.GetType().IsPrimitive || value.GetType().IsEnum || value is string || value is decimal [](start = 25, length = 92)

Please share the logic with other methods

333fred commented:
Don review pass (commit 41)

AlekseyTs commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplay.cs:
```diff
@@ -295,5 +286,146 @@ public static string FormatLiteral(char c, bool quote)
+            builder.Add(new SymbolDisplayPart(kind, null, valueString));
+        }
+
+        private static ObjectDisplayOptions ToObjectDisplayOptions(bool quoteStrings, bool useHexadecimalNumbers)
```
>ToObjectDisplayOptions [](start = 44, length = 22)

Can we share this helper across compilers?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/SymbolDisplay/SymbolDisplay.vb:
```diff
@@ -115,30 +115,121 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+
+            Dim builder = ArrayBuilder(Of SymbolDisplayPart).GetInstance()
+            If type Is Nothing Then
+                AddConstantValue(builder, value, ToObjectDisplayOptions(If(format, SymbolDisplayFormat.VisualBasicErrorMessageFormat).ConstantValueOptions))
```
>If(format, SymbolDisplayFormat.VisualBasicErrorMessageFormat) [](start = 72, length = 61)

Would it make sense to de-dupe this? Both branches calculate this unconditionally.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/SymbolDisplay/SymbolDisplay.vb:
```diff
@@ -115,30 +115,121 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+
+            Dim builder = ArrayBuilder(Of SymbolDisplayPart).GetInstance()
+            If type Is Nothing Then
+                AddConstantValue(builder, value, ToObjectDisplayOptions(If(format, SymbolDisplayFormat.VisualBasicErrorMessageFormat).ConstantValueOptions))
```
>If(format, SymbolDisplayFormat.VisualBasicErrorMessageFormat) [](start = 72, length = 61)

De-dupe?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/SymbolDisplay/SymbolDisplay.vb:
```diff
@@ -115,30 +115,121 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+
+            Dim type = value.GetType()
+
+            Select Case type
```
>type [](start = 24, length = 4)

Would it be more efficient to switch on ```Type.​Get​Type​Code()```?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/SymbolDisplay/SymbolDisplay.vb:
```diff
@@ -115,30 +115,121 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+                Case Else
+                    Dim valueString = ObjectDisplay.FormatPrimitive(value, options)
+                    Debug.Assert(valueString IsNot Nothing)
+                    Dim kind = If(type = GetType(Boolean), SymbolDisplayPartKind.Keyword, SymbolDisplayPartKind.NumericLiteral)
```
>type = GetType(Boolean) [](start = 34, length = 23)

VB has a ```TypeOf Is``` operation. Or perhaps use the same TypeCode value from previous checks.

AlekseyTs commented:
Done with review pass (iteration 41)

jcouv commented:
I've converted the PR to a draft (WIP). Please mark as ready for review when the PR is ready to move forward. Thanks

somendra007 commented:
I'm in future

Neme12 commented:
I'm really sorry for abandoning all these PRs, I've been struggling with severe mental health issues :(

CyrusNajmabadi commented:
Nothing to be sorry about. Take care of yourself first and foremost
